hbase-client/src/main/java/org/apache/hadoop/hbase/client/trace/TableOperationSpanBuilder.java

```diff
@@ -46,51 +44,61 @@
 import org.apache.yetus.audience.InterfaceAudience;
 
 /**
- * Construct {@link io.opentelemetry.api.trace.Span} instances originating from
- * "table operations" -- the verbs in our public API that interact with data in tables.
+ * Construct {@link Span} instances originating from "table operations" -- the verbs in our public
+ * API that interact with data in tables.
  */
 @InterfaceAudience.Private
-public class TableOperationSpanBuilder implements Supplier<Span> {
+public class TableOperationSpanBuilder<B extends TableOperationSpanBuilder<B>>
+  extends TableSpanBuilder<B> {
 
   // n.b. The results of this class are tested implicitly by way of the likes of
   // `TestAsyncTableTracing` and friends.
 
   private static final String unknown = "UNKNOWN";
 
-  private TableName tableName;
-  private final Map<AttributeKey<?>, Object> attributes = new HashMap<>();
+  public TableOperationSpanBuilder(final AsyncConnectionImpl conn) {
+    super(conn);
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public B self() {
```

Comment:
Why we need to have so complicated class hierarchy? A class has a generic type parameter which is just itself?

Reply:
I've pushed a change that replaces build inheritance with a mix-in style of builder. I don't really like it any better, let me know what you think.